File: typescript/ai-sloppy-paste/src/utils/queryParser.ts
import { normalizeTag } from "./tags";

/**
 * Represents a parsed search query with extracted operators and filters
 */
export interface ParsedQuery {
  /** Tag filters - snippet must have all these tags (hierarchical matching) */
  tags: string[];
  /** Negative tag filters - snippet must NOT have these tags */
  notTags: string[];
  /** Boolean "is" filters - favorite, archived, untagged */
  is: string[];
  /** Boolean "not" filters - negated favorite, archived, untagged */
  not: string[];
  /** Exact phrase matches - must appear as substring in title or content */
  exactPhrases: string[];
  /** Remaining fuzzy text - all words must appear somewhere */
  fuzzyText: string;
  /** Quick flag to check if any operators are present */
  hasOperators: boolean;
}

/**
 * Valid values for is:/not: operators
 */
const VALID_BOOLEAN_OPERATORS = ["favorite", "archived", "untagged"] as const;

/**
 * Parses a search query string into structured operators and filters.
 *
 * Supported operators:
 * - tag:work          - Must have tag (supports hierarchy)
 * - not:tag:personal  - Must NOT have tag
 * - is:favorite       - Must be favorite/archived/untagged
 * - not:archived      - Must NOT be archived/favorite/untagged
 * - "exact phrase"    - Must contain exact phrase (case-insensitive)
 * - remaining text    - Fuzzy search (all words must match somewhere)
 *
 * Examples:
 * - "tag:work api"                    → Has work tag AND contains "api"
 * - "not:archived python"             → Not archived AND contains "python"
 * - "tag:work is:favorite rest"       → Work tag AND favorite AND contains "rest"
 * - '"meeting notes" urgent'          → Contains "meeting notes" exactly AND "urgent"
 *
 * Error handling: Invalid or malformed operators are silently treated as fuzzy text.
 *
 * @param query - The search query string from the user
 * @returns ParsedQuery object with extracted operators and filters
 */
export function parseSearchQuery(query: string): ParsedQuery {
  const result: ParsedQuery = {
    tags: [],
    notTags: [],
    is: [],
    not: [],
    exactPhrases: [],
    fuzzyText: "",
    hasOperators: false,
  };

  // Empty query - return empty result
  if (!query.trim()) {
    return result;
  }

  // Step 1: Extract quoted phrases for exact matching
  // Match content between double quotes
  const quoteRegex = /"([^"]+)"/g;
  let match;
  while ((match = quoteRegex.exec(query)) !== null) {
    const phrase = match[1].trim();
    if (phrase) {
      result.exactPhrases.push(phrase);
    }
  }

  // Remove quoted phrases from query for operator parsing
  let remainingQuery = query.replace(quoteRegex, "").trim();

  // Step 2: Parse operators from remaining text
  // Split by whitespace and process each token
  const tokens = remainingQuery.split(/\s+/);
  const fuzzyTokens: string[] = [];

  for (const token of tokens) {
    if (!token) continue;

    let consumed = false;

    // Check for not:tag:value (must check before tag: to avoid false match)
    if (token.startsWith("not:tag:")) {
      const tagValue = token.substring("not:tag:".length);
      if (tagValue) {
        result.notTags.push(normalizeTag(tagValue));
        consumed = true;
      }
    }
    // Check for tag:value
    else if (token.startsWith("tag:")) {
      const tagValue = token.substring("tag:".length);
      if (tagValue) {
        result.tags.push(normalizeTag(tagValue));
        consumed = true;
      }
    }
    // Check for not:value (boolean operators)
    else if (token.startsWith("not:")) {
      const notValue = token.substring("not:".length);
      if (VALID_BOOLEAN_OPERATORS.includes(notValue as any)) {
        result.not.push(notValue);
        consumed = true;
      }
    }
    // Check for is:value (boolean operators)
    else if (token.startsWith("is:")) {
      const isValue = token.substring("is:".length);
      if (VALID_BOOLEAN_OPERATORS.includes(isValue as any)) {
        result.is.push(isValue);
        consumed = true;
      }
    }

    // If token wasn't consumed by an operator, add to fuzzy search
    if (!consumed) {
      fuzzyTokens.push(token);
    }
  }

  // Step 3: Build fuzzy text from remaining tokens
  result.fuzzyText = fuzzyTokens.join(" ").trim();

  // Step 4: Set hasOperators flag
  // Include fuzzyText so that simple searches also trigger applySearchFilters
  result.hasOperators =
    result.tags.length > 0 ||
    result.notTags.length > 0 ||
    result.is.length > 0 ||
    result.not.length > 0 ||
    result.exactPhrases.length > 0 ||
    result.fuzzyText.length > 0;

  return result;
}

/**
 * Helper to check if a parsed query is empty (no operators and no text)
 */
export function isEmptyQuery(query: ParsedQuery): boolean {
  return !query.hasOperators && !query.fuzzyText;
}

File: swift/sloppy-paste/Sources/SloppyCore/Search/QueryParser.swift
import Foundation

/// The `is:`/`not:` flags a query can test.
public enum BooleanFilter: String, Sendable, Hashable, CaseIterable {
    case favorite
    case bookmarked
    case archived
    case untagged
}

/// A search query split into operators and the remaining fuzzy text.
public struct ParsedQuery: Sendable, Hashable {
    /// Snippet must have all these tags (hierarchical matching).
    public var tags: [String] = []
    /// Snippet must have none of these tags.
    public var notTags: [String] = []
    /// Snippet title must carry all these `Prefix:` contexts.
    public var contexts: [String] = []
    /// Snippet title must carry none of these contexts.
    public var notContexts: [String] = []
    public var isFilters: [BooleanFilter] = []
    public var notFilters: [BooleanFilter] = []
    /// Case-insensitive substrings of the title or content.
    public var exactPhrases: [String] = []
    /// Remaining words; each must appear in the title, content or a tag.
    public var fuzzyText = ""

    public init(
        tags: [String] = [], notTags: [String] = [], contexts: [String] = [], notContexts: [String] = [],
        isFilters: [BooleanFilter] = [], notFilters: [BooleanFilter] = [], exactPhrases: [String] = [],
        fuzzyText: String = ""
    ) {
        self.tags = tags
        self.notTags = notTags
        self.contexts = contexts
        self.notContexts = notContexts
        self.isFilters = isFilters
        self.notFilters = notFilters
        self.exactPhrases = exactPhrases
        self.fuzzyText = fuzzyText
    }

    /// True when anything but plain fuzzy text is present.
    public var hasStructuredOperators: Bool {
        !tags.isEmpty || !notTags.isEmpty || !contexts.isEmpty || !notContexts.isEmpty
            || !isFilters.isEmpty || !notFilters.isEmpty || !exactPhrases.isEmpty
    }

    /// True when the query filters anything; fuzzy text counts.
    public var hasOperators: Bool {
        hasStructuredOperators || !fuzzyText.isEmpty
    }

    public var isEmpty: Bool { !hasOperators }
}

/// Parses the search field syntax:
///
/// - `tag:work` / `not:tag:personal`: has (or lacks) the tag or a child of it
/// - `ctx:asl` / `not:ctx:asl`: the title does (or does not) carry the `asl:` context
/// - `is:favorite` / `not:archived`: favorite, bookmarked, archived or untagged
/// - `"exact phrase"`: case-insensitive substring of the title or content
/// - anything else: fuzzy words
///
/// Malformed operators (`tag:`, `is:invalid`, an unclosed quote) fall back to fuzzy text.
public enum QueryParser {
    public static func parse(_ query: String) -> ParsedQuery {
        var result = ParsedQuery()
        guard !query.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else { return result }

        let quoted = /"([^"]+)"/
        for match in query.matches(of: quoted) {
            let phrase = String(match.1).trimmingCharacters(in: .whitespacesAndNewlines)
            if !phrase.isEmpty { result.exactPhrases.append(phrase) }
        }
        let remaining = query.replacing(quoted, with: "")

        var fuzzyTokens: [Substring] = []
        for token in remaining.split(whereSeparator: \.isWhitespace) {
            if !consume(token, into: &result) { fuzzyTokens.append(token) }
        }
        result.fuzzyText = fuzzyTokens.joined(separator: " ")
        return result
    }

    /// Records an operator token and returns whether it was one. `not:tag:` and
    /// `not:ctx:` are checked before the generic `not:`, which would otherwise
    /// reject them as unknown booleans and leave them as fuzzy text.
    private static func consume(_ token: Substring, into result: inout ParsedQuery) -> Bool {
        if let value = value(of: token, after: "not:tag:") {
            result.notTags.append(TagNormalization.normalizeTag(value))
        } else if let value = value(of: token, after: "not:ctx:") {
            result.notContexts.append(TitleContext.normalize(value))
        } else if let value = value(of: token, after: "tag:") {
            result.tags.append(TagNormalization.normalizeTag(value))
        } else if let value = value(of: token, after: "ctx:") {
            result.contexts.append(TitleContext.normalize(value))
        } else if let value = value(of: token, after: "not:"), let filter = BooleanFilter(rawValue: value) {
            result.notFilters.append(filter)
        } else if let value = value(of: token, after: "is:"), let filter = BooleanFilter(rawValue: value) {
            result.isFilters.append(filter)
        } else {
            return false
        }
        return true
    }

    /// The non-empty rest of `token` after `prefix`, or nil.
    private static func value(of token: Substring, after prefix: String) -> String? {
        guard token.hasPrefix(prefix), token.count > prefix.count else { return nil }
        return String(token.dropFirst(prefix.count))
    }
}
